### PR TITLE
get closure-type-checking running again

### PR DESCRIPTION
**Pull request by samthor**
@brendankenny 

This gets us up and running again, at least with warnings, not errors.

This works around [#954](https://github.com/google/closure-compiler/issues/954) in an interesting way, but I have some thoughts on fixing that properly too.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - "5.0"
+  - "5.1"
```
Comment (brendankenny) [inline .travis.yml]: why does this matter?

File: closure/closure-type-checking.js [hunk 2]
```diff
@@ -21,24 +21,32 @@ const closureCompiler = require('google-closure-compiler').gulp();
...
 
 /* eslint-disable camelcase */
 gulp.task('js-compile', function() {
   return gulp.src([
     'closure/typedefs/*.js',
-    'closure/third_party/*.js',
-    'audits/**/*.js',
-    'lib/icons.js',
-    'aggregators/**/*.js'
+    'src/closure/third_party/*.js',
```
Comment (brendankenny) [inline closure/closure-type-checking.js]: this shouldn't be under `src/`?

File: closure/closure-type-checking.js [hunk 3]
```diff
@@ -21,24 +21,32 @@ const closureCompiler = require('google-closure-compiler').gulp();
...
+    .pipe(replace('require(\'./formatter\');', '{};'))
+    .pipe(replace('require(\'../../../formatters/formatter\');', '{};'))
+
+    // Replace any non-local import (e.g. not starting with .) with a dummy type. These are likely
+    // the built-in Node modules. But not always, so TODO(samthor): Fix this.
+    .pipe(replace(/require\(\'[^\.].*?\'\)/g, '/** @type {*} */ ({})'))
 
     .pipe(closureCompiler({
+      externs: [
+        'closure/externs.js'
```
Comment (brendankenny) [inline closure/closure-type-checking.js]: since this file has an `@externs` annotation in it, you can just move this up to `gulp.src` with the rest, which allows us to keep everything in the same pipeline

File: closure/externs.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+/**
+ * @externs
+ */
+
+/* eslint-disable no-unused-vars */
+
+/** @type {string} */
+var __dirname;
```
Comment (brendankenny) [inline closure/externs.js]: I'd rather keep these under `third_party/` and `typedefs`, for externs for third-party and first-party code, respectively

File: closure/externs.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+/** @type {!Object} */
+var WebInspector;
```
Comment (brendankenny) [inline closure/externs.js]: were the externs in `closure/third_party/WebInspector.js` not working for this?

File: src/aggregators/aggregate.js [hunk 6]
```diff
@@ -201,7 +201,7 @@ class Aggregate {
       // TODO(paullewis): Remove once coming soon audits have landed.
       if (expected[e].comingSoon) {
         subItems.push({
-          value: String.raw`¯\_(ツ)_/¯`,
+          value: 'Coming Soon',  // TODO(samthor): Can't use String.raw, Closure has it defined incorrectly
```
Comment (brendankenny) [inline src/aggregators/aggregate.js]: isn't the shrugging guy on purpose? :) Can just go with '¯\\_(ツ)_/¯'

File: src/audits/performance/first-meaningful-paint.js [hunk 7]
```diff
@@ -234,8 +234,9 @@ module.exports = FirstMeaningfulPaint;
 
 /**
  * Math.max, but with NaN values removed
+ * @param {...number} numbers
  */
-function max() {
-  const args = [...arguments].filter(val => !isNaN(val));
+function max(...numbers) {
```
Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: we don't have rest parameters in node 5. Can you use the style guide suggestion of `@param {...number} var_args` but then still use `arguments` in the code?

File: src/audits/performance/speed-index-metric.js [hunk 8]
```diff
@@ -17,7 +17,7 @@
 
 'use strict';
 
-const speedline = require('speedline');
+const speedline = /** @type {function(*)} */ (require('speedline'));
```
Comment (brendankenny) [inline src/audits/performance/speed-index-metric.js]: :(

File: src/lib/traces/tracing-processor.js [hunk 9]
```diff
@@ -47,7 +47,7 @@ global.HTMLImportsLoader.hrefToAbsolutePath = function(path) {
 require('../../../third_party/traceviewer-js/');
 const traceviewer = global.tr;
 
-class TraceProcessor {
+class TracingProcessor {
```
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: any reason for this? match the file name?

File: closure/closure-type-checking.js [hunk 10]
```diff
@@ -21,24 +21,32 @@ const closureCompiler = require('google-closure-compiler').gulp();
...
+    'src/aggregators/**/*.js'
   ])
     // TODO: hack to remove `require`s that Closure currently can't resolve.
-    .pipe(replace(devtoolsRequire, ''))
+    .pipe(replace('require(\'../lib/traces/devtools-timeline-model\');', 'null;'))
     .pipe(replace('require(\'../../lib/web-inspector\').Color.parse;',
         'WebInspector.Color.parse;'))
+    .pipe(replace('require(\'../../lib/traces/tracing-processor\');', '/** @type {?} */ (null);'))
+    .pipe(replace('require(\'./formatter\');', '{};'))
+    .pipe(replace('require(\'../../../formatters/formatter\');', '{};'))
```
Comment (brendankenny) [inline closure/closure-type-checking.js]: this makes me sad, but I assume we can remove it as we add annotations to the formatters, etc?

File: src/audits/performance/speed-index-metric.js (hunk 8 above)
Comment (brendankenny) [inline src/audits/performance/speed-index-metric.js]: can we do this with the `gulp-replace`/externs approach instead? That way it keeps it all in one place (`closure/`).

**Comment by brendankenny** [comment]
reviewed! Current build failure should be fixed by not using rest parameters while we still support node 5.

File: .travis.yml (hunk 1 above)
Comment (brendankenny) [inline .travis.yml]: ah, if this was for the build failure, this won't help :) Rest parameters were added in 6

Comment (samthor) [inline .travis.yml]: Fixed this in the method.

File: closure/closure-type-checking.js (hunk 3 above)
Comment (samthor) [inline closure/closure-type-checking.js]: Done

File: closure/closure-type-checking.js (hunk 2 above)
Comment (samthor) [inline closure/closure-type-checking.js]: whoops, fixed.

File: closure/externs.js (hunk 5 above)
Comment (samthor) [inline closure/externs.js]: They are now.

File: src/audits/performance/first-meaningful-paint.js (hunk 7 above)
Comment (samthor) [inline src/audits/performance/first-meaningful-paint.js]: Fixed, I now use `_`, which Closure and ES are happy with (unused var)

File: src/lib/traces/tracing-processor.js (hunk 9 above)
Comment (samthor) [inline src/lib/traces/tracing-processor.js]: Yes, I thought it fixed something. Reverted.

File: src/aggregators/aggregate.js (hunk 6 above)
Comment (samthor) [inline src/aggregators/aggregate.js]: I've re-added with the extra \\

File: closure/externs.js (hunk 4 above)
Comment (samthor) [inline closure/externs.js]: Done

File: src/audits/performance/speed-index-metric.js (hunk 8 above)
Comment (samthor) [inline src/audits/performance/speed-index-metric.js]: Done

**Comment by samthor** [comment]
@brendankenny this should be good to go now. I've got more fixes in a follow-up patch.

**Comment by brendankenny** [comment]
nice job! LGTM

**Comment by youngdelaney** [comment]
Well don't know what i'm supposing comment on far this comment because i don't understand what you talking bout you not making plain to understand ok

File: src/aggregators/aggregate.js [hunk 11]
```diff
@@ -206,7 +206,7 @@ class Aggregate {
       // TODO(paullewis): Remove once coming soon audits have landed.
       if (expected[e].comingSoon) {
         subItems.push({
-          value: String.raw`¯\_(ツ)_/¯`,
+          value: '¯\\_(ツ)_/¯', // TODO(samthor): Patch going to Closure, String.raw is badly typed
```
Comment (paullewis) [inline src/aggregators/aggregate.js]: Noooooooooo!

Fine.

File: src/audits/performance/critical-request-chains.js [hunk 12]
```diff
@@ -63,7 +63,7 @@ class CriticalRequestChains extends Audit {
    */
   static audit(artifacts) {
     let chainCount = 0;
-    function walk(node, depth, startTime) {
+    function walk(node, depth) {
```
Comment (paullewis) [inline src/audits/performance/critical-request-chains.js]: This change seems unrelated to the PR.

Comment (paullewis) [inline src/audits/performance/critical-request-chains.js]: Mind you, it's fine actually... the fix needs to exist. Go for it.

Comment (samthor) [inline src/audits/performance/critical-request-chains.js]: Yes, Closure fails without this fix. That's one of the points of this
change :)
